### DISCUSSION TITLE
Woocommerce plugin update (Greenfield)

### DISCUSSION THREAD

**Discussion by astupidmoose** [btcpayserver/btcpayserver#2142] [Plugins & integrations]
As the new Greenfield API continues to grow, we should probably consider rewriting the woocommerce plugin to interact with it vs the old Bitpay API. 

I'm going to start this discussion as a starting point. I think we should discuss core features and possibly try to gain some dev support for this.

There is a woocommerce plugin repo, but discussions are not enabled there. 

Personally I'd like to see a native modal invoice option. This keeps your customer on the order screen as they pay.

Another user suggested an integration with the new API to use woocommerce refunds with btcpay payments.

**Comment by Kukks** [comment]
IMO, it's better to have one central repo for these discussions.

@dennisreimann has just finished a PR that fixes custom order handling in WooCommerce so that it will work no matter what.

@ndeet will be working on some really great features for the plugin in January which I am sure you will love @astupidmoose. The feature focuses on allowing you to present the payment methods in btcpay as individual payment methods on your checkout. This allows you to integrate with other WooCommerce plugins in a more fine grained way (eg 5% off if you use Lightning)

After this, the plan is to start a big push from my end to:
* Create libraries for many programming languages for Greenfield API
* Create new plugin integrations that use the new Greenfield API. I would avoid upgrading the plugins and start on a new repository and plugin to avoid dealing with migrations and any fatal downtime from an accidental automatic update.

For WooCommerce, the plan will be to create a new, client repository, not based on some 7year old bitpay codebase and build it with the best experience we can come up with.

**Comment by pavlenex** [comment]
This is an interesting idea, it would also prevent merchants from having to create a-record domains https://twitter.com/lukechilds/status/1350020207318769665 and enable people to run btcpay through Tor where the Woo would have it's own checkout and just ping the API.

**Reply by dennisreimann** [reply]
At some point WooCommerce and BTCPay Server must talk though and that's where in this case the API would need to be called via Tor, because the Umbrel server in this example isn't exposed via clearnet. This means that the server hosting the WooCommerce shop would need to support communicating via Tor as well.
I'm sure this can be done, just pointing it out because this won't be the plug and play method a regular PHP host provides.

**Reply by pavlenex** [reply]
What I also like about this solution is the idea of having an integrated checkout that doesn't redirect to btcpay server. It can be a big deal if we continue to have problems with Google's flagging or simply a malicious regulators trying to push regulations on merchants who use btcpay (just an example), this way we can have whitelabel option in the plugin that would just use integrated checkout, and would help us test our API better :smile:

**Comment by ndeet** [comment]
> After this, the plan is to start a big push from my end to:
> - Create libraries for many programming languages for Greenfield API

Are you thinking about using some approach like LND where they wrote everything in .proto standard and then you can generate clients for every language from it? It works ok for LND but the generated clients are really ugly and have a lot of useless boilerplate (at least for the generated PHP code) but it has its benefits. 

> - Create new plugin integrations that use the new Greenfield API. I would avoid upgrading the plugins and start on a new repository and plugin to avoid dealing with migrations and any fatal downtime from an accidental automatic update.
> For WooCommerce, the plan will be to create a new, client repository, not based on some 7year old bitpay codebase and build it with the best experience we can come up with.

Yes this is a must imo, the bitpay code is a mess and a lot can be thrown out/streamlined and will need almost a full refactor anyway because of the different auth and structure of Greenfield API; trying making old and new API work in the existing plugin will lead to. a lot of pain, imo

**Comment by pavlenex** [comment]
Adding Data Retention requested by @astupidmoose 

https://github.com/btcpayserver/btcpayserver/discussions/2665#discussioncomment-971654
